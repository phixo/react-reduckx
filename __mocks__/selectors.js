import { createSelector } from 'reselect';

const branchSelector = state => state.mock;

const mockDataSelector = createSelector(
    branchSelector,
    state => state && state.data
);

const mockCounterSelector = createSelector(
    branchSelector,
    state => state && state.counter
);

export { branchSelector, mockCounterSelector, mockDataSelector };
