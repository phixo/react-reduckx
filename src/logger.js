/* eslint-disable no-console */
const COLOR_ACTION = '#03A9F4';
const COLOR_NEXT = '#4CAF50';
const COLOR_PREV = '#9E9E9E';

const Logger = {
    log: (action = {}, prevState = {}, nextState = {}) => {
        if (action.type) {
            console.group(action.type);
            console.log('%cprev state', `color: ${COLOR_PREV};`, prevState);
            console.log('%caction    ', `color: ${COLOR_ACTION};`, action);
            console.log('%cnext state', `color: ${COLOR_NEXT};`, nextState);
            console.groupEnd();
        }
    },
};

export default Logger;
