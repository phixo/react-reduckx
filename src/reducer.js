import Logger from './logger';

const combineReducers = reducers => {
    const branches = Object.keys(reducers);

    return (state = {}, action) => {
        const nextState = {};

        branches.forEach(branch => {
            const reduceFn = reducers[branch];
            const prevState = state[branch];

            nextState[branch] = reduceFn(prevState, action);
        });

        Logger.log(action, state, nextState);

        return nextState;
    };
};

const createReducer = (initialState, actionsMap) => (state, action) => {
    const prevState = state || initialState;
    const reduceFn = actionsMap[action.type || ''];

    if (!reduceFn) {
        return prevState;
    }

    return {
        ...prevState,
        ...reduceFn(prevState, action),
    };
};

const reduceWith = (reduceFn, state, action) => {
    if (!reduceFn) {
        return state;
    }
    return reduceFn(state, action);
};

export { combineReducers, createReducer, reduceWith };
