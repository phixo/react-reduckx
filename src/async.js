import { reduceWith } from './reducer';

const decreasePendingState = state => ({
    isPending: Math.max(0, state.isPending - 1),
});

const increasePendingState = state => ({ isPending: state.isPending + 1 });

const createAsyncReducer = (reduceFn, isPending) => (state, action) => ({
    ...reduceWith(reduceFn, state, action),
    ...(isPending ? increasePendingState(state) : decreasePendingState(state)),
});

const createAsyncAction = (type, operation, defaultOptions) => {
    return (options = defaultOptions) => ({ getState, dispatch }) => {
        dispatch({
            type: `${type}_PENDING`,
            options,
        });

        const promise = new Promise(async (resolve, reject) => {
            try {
                const res = await operation(options, getState, dispatch);
                const payload = res.data;

                dispatch({
                    type: `${type}_SUCCESS`,
                    payload,
                    options,
                });

                resolve(payload);
            } catch (error) {
                dispatch({
                    type: `${type}_FAIL`,
                    payload: error,
                    options,
                });

                reject(error);
            }
        });

        return promise;
    };
};

const createAsyncActionReducers = (type, options) => {
    const { onPending, onSuccess, onFail } = options;

    return {
        [`${type}_PENDING`]: createAsyncReducer(onPending, true),
        [`${type}_SUCCESS`]: createAsyncReducer(onSuccess, false),
        [`${type}_FAIL`]: createAsyncReducer(onFail, false),
    };
};

export { createAsyncAction, createAsyncActionReducers };
