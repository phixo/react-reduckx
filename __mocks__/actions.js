import axios from 'axios';
import { createAsyncAction } from '../src';
import { ACTION_ASYNC, ACTION_SYNC } from './action-types';

const mockAsyncActionSuccess = createAsyncAction(ACTION_ASYNC, id =>
    axios.get(`/mock/${id}`)
);

const mockAsyncActionFail = createAsyncAction(ACTION_ASYNC, () =>
    axios.get(`/mock/fail`)
);

const mockSyncAction = payload => ({
    type: ACTION_SYNC,
    payload,
});

export { mockAsyncActionSuccess, mockAsyncActionFail, mockSyncAction };
