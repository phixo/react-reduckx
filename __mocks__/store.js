import { combineReducers, createStore } from '../src';
import mockReducer, { initialState as mockState } from './reducers';

const rootReducer = combineReducers({
    mock: mockReducer,
    foo: mockReducer,
});

const initialState = {
    mock: { ...mockState },
};

export const { ReduxProvider, useRedux } = createStore(
    rootReducer,
    initialState
);
