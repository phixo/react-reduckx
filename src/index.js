import createStore from './store';
import { createAsyncAction, createAsyncActionReducers } from './async';
import { combineReducers, createReducer, reduceWith } from './reducer';

const Reduckx = {
    combineReducers,
    createAsyncAction,
    createAsyncActionReducers,
    createReducer,
    createStore,
    reduceWith,
};

export {
    Reduckx as default,
    combineReducers,
    createAsyncAction,
    createAsyncActionReducers,
    createReducer,
    createStore,
    reduceWith,
};
