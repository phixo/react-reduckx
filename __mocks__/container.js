import React, { useEffect } from 'react';
import { useRedux } from './store';

import { mapState, actionsMap } from './defaults';

const MockContainer = () => {
    const [{ counter, data }, { syncFn, asyncFn }] = useRedux(
        mapState,
        actionsMap
    );

    useEffect(() => {
        try {
            syncFn(1);
            asyncFn(1);
        } catch (error) {
            throw Error(error);
        }
    }, []);

    return (
        <>
            {counter} {JSON.stringify(data)}
        </>
    );
};

export default MockContainer;
