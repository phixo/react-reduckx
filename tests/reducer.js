import { combineReducers, reduceWith } from '../src';
import { mockSyncAction } from './__mocks__/actions';
import mockReducer, { initialState } from './__mocks__/reducers';

const test = {
    reduceFn: jest.fn(),
};

const state = {
    name: 'John Doe',
};

const action = mockSyncAction({ id: 1 });

describe('Reducer', () => {
    const spy = jest.spyOn(test, 'reduceFn');

    afterEach(() => {
        spy.mockClear();
    });

    it('Should combine reducers', () => {
        const reducer = combineReducers({
            mock: mockReducer,
            foo: mockReducer,
        });
        const nextState = reducer(initialState, { type: 'MOCK/ACTION_SYNC' });
        expect(nextState).toMatchSnapshot();
    });

    it('Should wrap reducer', () => {
        reduceWith(test.reduceFn, state, action);

        expect(spy).toHaveBeenCalledWith(state, action);
    });
});
