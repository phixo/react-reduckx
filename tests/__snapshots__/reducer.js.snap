// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer Should combine reducers 1`] = `
Object {
  "foo": Object {
    "counter": NaN,
    "data": Object {},
    "isPending": 0,
  },
  "mock": Object {
    "counter": NaN,
    "data": Object {},
    "isPending": 0,
  },
}
`;
